### DISCUSSION TITLE
GC and lifetime of objects captured by closure

### DISCUSSION THREAD

**Discussion by acaly** [dotnet/csharplang#4605] [Q&A]
I noticed this question recently: How does lambda affects the exact lifetime of its captured objects?

Yes it should keep it alive in order to have access to it. But consider the code below:
```c#
using System;
using System.Runtime.CompilerServices;

public class C {
    public int Value;
    public Func<C, int> M(C x, C y) {
        M2(z => z.Value == y.Value);
        return z => z.Value + x.Value;
    }
    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void M2(Func<C, bool> x) {
    }
}
```

Here `C y` is only temporarily used in `M`, but its reference is carried by the closure that returns, as confirmed [here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMA6AJQFcA7GASzDlIGEB7MABzoBtEBlRADc6AYzhQA3LlwoAzMTSFmhAN65CG4vLq1CANQjdKcKTk1biAVgA8zADSEdMAHyEAsgAplADwfKAngCUqurmmm5oHgBehAC8rlGkBkZwcbGE/kmGxoGmYZooAOyEMfElWSmEANSE3hXGeZoAvqEaANpucDAAFqwAJgCSnNwenT39QxzcAPIc9KzUUKQAcqwD1Nw6OgDmgQC6rRYkAGzEACzukSg29oTArKzcrt7BambNuE1AA). If the caller to `M` holds a reference to the returned delegate, both `x` and `y` would be alive, although no one should be able to access `y` any more. 

Isn't this considered to be a kind of memory leak? I imagine the problem will deteriorate for larger methods that use many lambdas (for example, for Linq). 

Note that the spec says:

> However, the lifetime of a captured outer variable is extended ***at least*** until the delegate or expression tree created from the anonymous function becomes eligible for garbage collection.

So the compiler seems to be obeying the spec.

**Comment by qrli** [comment]
As a default implementation, the closure is allocated on heap, so it is up to GC to clean them up, just as all reference objects. If you count that as a leak, then all reference objects also leak, before GC kicks in.

Surely the compiler can do better (to not allocate on heap) in theory, but it is hard if the M2 is in another scope, and impossible if M2 is in another assembly, because the compiler need to prove that M2 does not store the reference to the lambda. That's one reason why local method feature was introduced, which makes it easier to do that optimization.

**Reply by acaly** [reply]
Maybe I was not very clear. Let me explain myself again.

I was not asking the compiler to not allocate heap objects. I was asking the compiler to make 2 different heap objects, one holding `x`, the other holding `y`.

> If you count that as a leak, then all reference objects also leak, before GC kicks in.

The correctness of GC is to collect everything that is not accessible any more. If we use this criteria, then definitely there is a leak, because it's possible that the lambda that captures `y` is collected and the lambda that captures `x` is not (or never). In that period of time, no matter how hard the GC tries, it's impossible to collect the object passed in as `y`, because it's hold by a lambda that never accesses it. 

In general, I don't think the language/library is allowed to make such patterns possible. For example, `List<T>` must clear its array upon removing an object (for example, removing the last element). Because this object will never be accessible by anyone, the `List<T>` must ensure GC can collect it if it tries to. However, as I said, here the lambda is keeping something inaccessible to be alive and survive from GC.

**Reply by HaloFour** [reply]
You're trading slightly longer lifetimes for a guarantee of more allocations.  I'd expect that the latter would be much more expensive in the common cases.

**Reply by acaly** [reply]
@HaloFour Yes, I would agree that in most cases the current behavior is more efficient. But in some cases, that is not **slightly longer** lifetime but may be much much longer. In some UI framework, it's very common for the entire UI window and all the resources to be kept alive by a single delegate. I still think the correctness of GC should never be compromised.

**Reply by acaly** [reply]
@HaloFour I think in this case, maybe it's possible for roslyn to warn programmer that it's generating some code that shares the storage for different lambdas with possibly different lifetime. At least for myself, I wasn't aware of this problem before. 

What do you think?

**Reply by HaloFour** [reply]
That's something you can correct for in your code, reduce the scopes that you are capturing in shared delegates.  Use local functions, etc.

Having the compiler generate separate closure classes for every capturing delegate within a method would be substantially less efficient.  In a method with a chain of LINQ expressions you're talking potentially ten times more allocations.  And that still couldn't guarantee that the lifetime of a captured variable would be limited to the scope in which it was captured as if any of the delegates share captured variables they would have to be in the same closure class.

**Reply by HaloFour** [reply]
> I think in this case, maybe it's possible for roslyn to warn programmer that it's generating some code that shares the storage for different lambdas with possibly different lifetime. At least for myself, I wasn't aware of this problem before.
>
> What do you think?

Sounds like good fodder for an analyzer.  It might be a challenge to ascertain that certain delegates might have a longer lifetime, but you could probably start with any delegate returned by the method.

**Reply by acaly** [reply]
As a reference, I would mention Lua. Lua is handling this correctly, as it create one heap object (upvalue) for each captured variable and won't share at all. Obviously that's the effort trying to avoid the exact problem I described. To reduce the overhead of creating too many upvalues, Lua would only create them when creating the closure, instead of when entering the block.

Of course Lua doesn't target at high performance by any means like CLR. So it make sense if both uses different approaches.

Also I don't think most of the LINQ expression would capture any variable. They are mostly operating on the parameters provided, although there are many exceptions.

**Reply by qrli** [reply]
Note that it is not specific to closure which may extend life time of objects. Any object can, even a local variable can as long as the it is kept in scope, e.g. by a long or infinite loop. Nothing makes closure special. It is the nature of GC.

In case the optimization is really necessary, you can always optimize your code to reduce capture or scope.
But in majority cases, GC cost is has a much bigger impact, and more allocations is just a bad choice.

**Reply by acaly** [reply]
> Nothing makes closure special.

No. Keeping something inaccessible alive is special. Sometimes it's even necessary to trace the reference graph to find why a specific object isn't properly collected. This can be hard. I once ran into this situation when I was trying to make an editor with plugin system that loads and unloads dlls. In CLR, an assembly won't be unloaded unless nothing holds references to it. Because the plugin file can be large (when containing manifest resources), if we are careless about unloading the assembly, the application can easily eat up system memory. I can imagine how frustrated one can be if spent long time on a memory leak which is actually caused by compiler.

A long or infinite loop can keep whatever it uses alive, because it needs to access the object. Conversely, the closure in the example above should not keep the object alive because it does not access the object. Both are the correctness of the GC. We usually say GC is indeterministic and unpredictable, but that's about when. Given enough time (or by calling `GC.Collect` for enough times), the behavior of the GC should be deterministic and predictable in that it should collect every dead object.

To reiterate, this is not about to optimize the program or to use less memory. It's about to ensure that the object can be collected once it's no longer used.

**Reply by CyrusNajmabadi** [reply]
>  It's about to ensure that the object can be collected once it's no longer used.

The definition of 'no longer used' is not well defined.  If i have a class and i assign a field to it that i never read later, teh same issue will occur.  The gc will not collect the object the field points at.  The GC doesn't collect resources that are no longer used.  It collects resources that are no longer reachable.  This is an important distinction.  So, as long as the object is reachable this can happen.

This has been a known tradeoff we've made *in the compiler* (not the language) for a very long time.  The approach taken here has this known downside, weighed against other very positive upsides.

**Reply by acaly** [reply]
> The definition of 'no longer used' is not well defined. 

It's well-defined by CLR, on which GC is made, but not by C#, which is more complicated and vague as there can be compiler generated types. I never blamed the GC on not collecting the object. I said this is a problem (or _tradeoff_) of C#, and that's why I'm asking it here, not in coreclr repo.

> The approach taken here has this known downside, weighed against other very positive upsides.

This is what I was hoping to know. Thank you for confirming.

**Reply by CyrusNajmabadi** [reply]
> It's well-defined by CLR,

Right, but the definition is not "once it's no longer used".  :)

**Reply by TonyValenti** [reply]
Wow!  It seems a lot of people are missing this point!

The observation from the original author is that, when a developer creates a closure, the C# compiler may generate source code that captures more variables/values/objects than the developer expressed.

Seeing this posted here, I'm pretty confident I've been a victim of this "feature" in the past.  It is my (strong) preference that closures only capture items that are expressed and aren't free to also capture other variables just because it feels like it.

**Reply by HaloFour** [reply]
The compiler will only capture the locals you actually enclose, nothing more.  But if you create multiple closures within the same method then the compiler will create a single shared closure with all of the enclosed locals.  This is to enable all of the closures to share a single state which is observable between the different delegates regardless of the order in which they are invoked or on which threads.  If you have multiple delegates in a method and enclose different combinations of locals you can expect that the lifetime of those locals is as long as the lifetime of the longest lived delegate.

**Reply by CyrusNajmabadi** [reply]
> The observation from the original author is that, when a developer creates a closure, the C# compiler may generate source code that captures more variables/values/objects than the developer expressed.

That's not accurate.  What the original author wants is that once they feel that the value is no longer needed, that it be eligible for GC.  We do not capture more variables than the code expressed.  We capture all the variables that need to be captured.  

>  It is my (strong) preference that closures only capture items that are expressed 

That is how it works.

> and aren't free to also capture other variables just because it feels like it.

We do not do that.

The issue here is that once captured, it may be that the lifetime of the capture lasts longer than desirable.  This is a known decision that came about between balancing many desires, all of which are heavily impactful.

**Reply by bernd5** [reply]
@Cyrus good answer
Theoretically a compiler would be allowed to optimize the call to `M2` (including the lambda/delegate creation) completely away because it has no observable side-effects.
But because this is practically not always / easy detectable a compiler is allowed to not optimize it at all. This is especially true in cases where the called code is not known at compile time (virtual methods, interfaces...).

**Reply by acaly** [reply]
@CyrusNajmabadi Actually, @TonyValenti accurately expressed my concern. The GC problem is just an (only) observable effect of the unexpected capture.

> We do not capture more variables than the code expressed. We capture all the variables that need to be captured.

The 2 lambdas as a whole do not capture more locals, but each of them captures one unnecessary local that is not accessed. If one of the lambda is garbage collected first, then since that time, the other lambda is capturing one more variable.

> This is to enable all of the closures to share a single state which is observable between the different delegates regardless of the order in which they are invoked or on which threads. 

Yes, the point of closure is that. But, as you can see, the 2 lambdas in the example don't communicate with each other and the behavior of the compiler is purely to avoid one more allocation. It's more like you write the code:

```c#
var a = new A();
var b = new B();
```

and the compiler actually generates a new class AB that contains both A and B and let the CLR to treat them as a single object (although this is impossible in the current CLR as you wouldn't be able to get a reference pointing to the middle of a object).

**Reply by CyrusNajmabadi** [reply]
> but each of them captures one unnecessary local that is not accessed.

I never said otherwise.  :)

Indeed, that was my point about talking about how you would observe the same issue if you just made a class and assigned these values to fields.  The GC doesn't have any concept of "this won't be accessed later, so i will cleanup".  It just has the concept of "these objects are unreachable, and thus eligible for collection."

> and the behavior of the compiler is purely to avoid one more allocation. 

Yes.  That's been stated many times that that was an *informed* and *intentional* design decision here.  There were several options, and this was chosen as the most suitable for dealing with the space.  I get that the cons of the approach we took were suboptimal for your case.  But that was known, evaluated, and decided on as acceptable versus the alternatives.  Our suggestion is to restructure code in this case to either avoid the need for larger capture objects, or to have your code explicitly inform the gc when objects are no longer reachable (i.e. by nulling them out in the closure).

**Reply by HaloFour** [reply]
@acaly 

> But, as you can see, the 2 lambdas in the example don't communicate with each other and the behavior of the compiler is purely to avoid one more allocation.

No, it's to avoid _x_ more allocations.  You'd need a separate allocation for each permutation of enclosed locals if you wanted to ensure that they can be collected individually.

The compiler intentionally optimizes for the common case here, that all of the delegates can be treated to have a common (and relatively short) lifetime.  Doing anything differently here might help in the niche case but would make that common case substantially worse.

**Reply by acaly** [reply]
> have your code explicitly inform the gc when objects are no longer reachable (i.e. by nulling them out in the closure).

That is a good point! At least we have something to do if we do end up with that pattern.

> No, it's to avoid x more allocations. You'd need a separate allocation for each permutation of enclosed locals

The number of allocations needed cannot exceed the number of local variables captured. As I said, Lua is already doing this. This is a O(N) vs O(1) cost, not a permutation.

Also, I am not saying we should change the current behavior, but it might not be that bad to have one more allocation. I remember that awaiting a synchronously completed ValueTask can even be slightly slower than awaiting Task in some cases. CLR GC is actually quite efficient at GEN0.

**Reply by HaloFour** [reply]
@acaly 

> The number of allocations needed cannot exceed the number of local variables captured. As I said, Lua is already doing this. This is a O(N) vs O(1) cost, not a permutation.

.NET is not Lua.  Each capturing delegate would need a separate allocation to enclose all of the locals used by that delegate.  Each combination of those enclosed variables would need a separate closure class and allocation to ensure that the lifetimes of the enclosed variables are not entangled with one another.  For 3 captured locals you'd need up to 7 different closure classes allocated.

**Reply by acaly** [reply]
Indeed. I didn't count the object allocated for each lambda. CLR can only have one this pointer for a delegate, which also needs to be allocated, while Lua can have many.

**Reply by HaloFour** [reply]
The C# team was briefly considering capture lists for lambdas: https://github.com/dotnet/roslyn/issues/117

Looks like it never really got past being considered worth the effort to try to figure out a non-horrible syntax.  And I don't know that the conversation ever got to the point of considering different closure allocations based on the capture lists which would have solved the GC issue in these cases.

**Reply by bernd5** [reply]
Is it anywhere explicitly defined how capturing has to be implemented in the spec?
I think this issue should be moved to roslyn/implementation as an optimization idea.

**Reply by acaly** [reply]
@bernd5 I mentioned the spec in my original post.

https://github.com/dotnet/csharplang/blob/main/spec/expressions.md#captured-outer-variables
> However, the lifetime of a captured outer variable is extended at least until the delegate or expression tree created from the anonymous function becomes eligible for garbage collection.

HaloFour's link to the other issue actually contains a real-world example: https://github.com/dotnet/roslyn/issues/117#issuecomment-319306294. Unfortunately, even that did not change LDM's dicision.

I guess the point is, normally one would just write lambdas in simple form, so unless the user actually realizing the problem, the new syntax does not help either. But after the user realizes it, without the new syntax, there are still something we can do to avoid it (for example as mentioned in the other issue, make a new block scope for that object).

**Reply by bernd5** [reply]
But a implementation which checks for non-overlapping capturing is not prohibited by that spec. It says only that the lifetime of captured object(s) are not bound to the current stackframe - their lifetime is extended to the lifetime of the capturing delegate / anonymous function / expressions tree / local function.
Roslyn does not go this way to reduce allocations and maybe because of additional reasons.

**Reply by TonyValenti** [reply]
Is there a way for me to identify functions that have multiple closures in them which could be causing this scenario?  I'm pretty sure I have a memory leak in my code because of a small variable that is inadvertently holding onto a large one.

**Reply by acaly** [reply]
Although it sounds like an analyzer's work, I think it's even easier to directly search for compiler generated closure types with more than one methods. This can be done with System.Reflection or Mono.Cecil.

**Reply by qrli** [reply]
@TonyValenti If you have VisualStudio Enterprise, it comes with nice memory debugger/analyzer, which can trace and diagnose the object reference chain easily.
If not, you can use WinDbg. There are a few low level SOS commands to learn, but as long as you get past that, things are pretty easy too. It typically takes me 15 minutes to search and relearn the commands to find the root cause each time.
